totalspineseg/utils/iterative_label.py

```diff
@@ -718,7 +689,287 @@ def iterative_label(
+    mask_labeled_z_indexes = [_[-1] for _ in ndi.center_of_mass(canonical_mask_labeled != 0, canonical_mask_labeled, sorted_labels)]
+
+    # Sort the labels by their z-index (reversed to go from superior to inferior)
+    sorted_z_indexes, sorted_labels = zip(*sorted(zip(mask_labeled_z_indexes, sorted_labels))[::-1])
+
+    return mask_labeled, num_labels, list(sorted_labels), list(sorted_z_indexes)
+
+def _merge_vertebrae_labels_with_no_disc_between(
+        seg,
+        mask_labeled,
+        num_labels,
+        sorted_labels,
+        sorted_z_indexes,
+        disc_sorted_z_indexes,
+    ):
+    '''
+    Combine sequential vertebrae labels if there is no disc between them.
+    '''
+    if num_labels == 0 or len(disc_sorted_z_indexes) == 0:
+        return mask_labeled, num_labels, sorted_labels, sorted_z_indexes
+
+    new_sorted_labels = []
+
+    # Store the previous label and the z index of the previous label
+    prev_l, prev_z = 0, 0
+
+    for l, z in zip(sorted_labels, sorted_z_indexes):
+        # Do not combine first and last vertebrae since it can be C1 or only contain the spinous process
+        if l not in sorted_labels[:2] and l != sorted_labels[-1] and prev_l > 0 and not any(z < _ < prev_z for _ in disc_sorted_z_indexes):
+            # Combine the current label with the previous label
+            mask_labeled[mask_labeled == l] = prev_l
+            num_labels -= 1
+
+        else:
+            # Add the current label to the new sorted labels
+            new_sorted_labels.append(l)
+            prev_l, prev_z = l, z
+
+    sorted_labels = new_sorted_labels
+
+    # Reduce size of mask_labeled
+    if mask_labeled.max() < np.iinfo(np.uint8).max:
+        mask_labeled = mask_labeled.astype(np.uint8)
+    elif mask_labeled.max() < np.iinfo(np.uint16).max:
+        mask_labeled = mask_labeled.astype(np.uint16)
+
+    # Get the z index of the center of mass for each label
+    canonical_mask_labeled = np.asanyarray(nib.as_closest_canonical(nib.Nifti1Image(mask_labeled, seg.affine, seg.header)).dataobj).round().astype(mask_labeled.dtype)
+    mask_labeled_z_indexes = [_[-1] for _ in ndi.center_of_mass(canonical_mask_labeled != 0, canonical_mask_labeled, sorted_labels)]
+
+    # Sort the labels by their z-index (reversed to go from superior to inferior)
+    sorted_z_indexes, sorted_labels = zip(*sorted(zip(mask_labeled_z_indexes, sorted_labels))[::-1])
+
+    return mask_labeled, num_labels, list(sorted_labels), list(sorted_z_indexes)
+
+def _merge_extra_labels_with_adjacent_vertebrae(
+        seg,
+        mask_labeled,
+        num_labels,
+        sorted_labels,
+        sorted_z_indexes,
+        extra_labels,
+    ):
+    '''
+    Combine extra labels with adjacent vertebrae labels.
+    This is useful for combining remaining of general vertebrae labels that introduce for region based training but not used in the final segmentation.
```

Comment:
Is this ok?
This is useful to combine segmentations of vertebrae introduced during region-based training, where the model sometimes outputs a general vertebrae label instead of the specific odd or even vertebrae. The process adjusts these remnants by merging them with the closest odd or even vertebrae to ensure correct segmentation.
https://github.com/neuropoly/totalspineseg/pull/56/commits/6b9f9148e6b22be7060be8ff8dc3ec0695ed574b